src/bindings/tsd.ts

```diff
@@ -0,0 +1,387 @@
+        if (seenObjectTypes.has(clazz)) {
+          typeName = assert(seenObjectTypes.get(clazz));
+          sb.push(typeName);
+          if (this.isPlainObject(clazz)) {
+            sb.push(mode == Mode.EXPORT ? "<never>" : "<undefined>");
+          }
+        } else {
+          if (this.isPlainObject(clazz)) {
+            typeName = "__Record" + clazz.id.toString();
+            sb.push(typeName);
+            sb.push(mode == Mode.EXPORT ? "<never>" : "<undefined>");
+            this.deferredTypings.push(this.makeRecordType(clazz, mode));
+          } else {
+            typeName = "__Internref" + clazz.id.toString();
+            sb.push(typeName);
+            this.deferredTypings.push(this.makeInternrefType(clazz));
+          }
+          seenObjectTypes.set(clazz, typeName);
+        }
+      }
+      if (type.is(TypeFlags.NULLABLE)) {
+        sb.push(" | null");
+      }
+      return sb.join("");
+    } else {
+      if (type == Type.bool) {
+        return "boolean";
+      }
+      if (type == Type.void) {
+        return "void";
+      }
+      if (type.isNumericValue) {
+        if (type.isLongIntegerValue) {
+          return "bigint";
+        }
+        return "number";
+      }
+    }
+    return "unknown";
+  }
+
+  makeRecordType(clazz: Class, mode: Mode): string {
+    var sb = new Array<string>();
+    var members = clazz.members;
+    sb.push("/** ");
+    sb.push(clazz.internalName);
+    sb.push(" */\ndeclare interface __Record");
+    sb.push(clazz.id.toString());
+    sb.push("<TOmittable> {\n");
+    if (members) {
+      for (let _keys = Map_keys(members), i = 0, k = _keys.length; i < k; ++i) {
+        let memberName = _keys[i];
+        let member = assert(members.get(memberName));
+        if (member.kind != ElementKind.FIELD) continue;
+        let field = <Field>member;
+        sb.push("  /** @type `");
+        sb.push(field.type.toString());
+        sb.push("` */\n  ");
+        sb.push(field.name);
+        sb.push(": ");
+        sb.push(this.toTypeScriptType(field.type, mode));
+        if (this.fieldAcceptsUndefined(field.type)) {
+          sb.push(" | TOmittable");
+        }
+        sb.push(";\n");
+      }
+    }
+    sb.push("}\n");
+    return sb.join("");
+  }
+
+  fieldAcceptsUndefined(type: Type): bool {
+    if (type.isInternalReference) {
+      return type.is(TypeFlags.NULLABLE);
+    }
+    return true;
+  }
+
+  makeInternrefType(clazz: Class): string {
+    var sb = new Array<string>();
+    sb.push("/** ");
+    sb.push(clazz.internalName);
+    sb.push(" */\n");
+    sb.push("declare class __Internref");
+    sb.push(clazz.id.toString());
+    sb.push(" extends Number {\n");
+    var base: Class | null = clazz;
+    do {
+      sb.push("  private __nominal");
+      sb.push(base.id.toString());
+      sb.push(": symbol;\n");
+      base = base.base;
+    } while (base);
+    sb.push("}\n");
+    return sb.join("");
+  }
+}
+
+// Helpers
+
+enum Mode {
+  IMPORT,
+  EXPORT
+}
```

Comment:
Is there a more readable way to do this, perhaps as template strings instead of concatenating?

Reply:
I assume you refer to the whole file? If so, both the .js and .d.ts generation utilize a mix as I saw fit. In particular it uses `.push`es where there are branches in between, and there are a lot. Looks a bit overloaded, I'd agree.